doc/arch/008-screen-architecture-v3.md

```diff
@@ -0,0 +1,759 @@
+# ADR 008: Screen Architecture (v3)
+
+## Status
+Accepted (Supersedes [001](./001-screen-controllers.md), [007](./007-screen-architecture-v2.md)) on 2019-10-01
+
+## Context
+Our [current screen architecture](./007-screen-architecture-v2.md) was a step above the [v1 screen architecture](./001-screen-controllers.md). However, it still has problems of its own:
+
+- The `UiStateProducer` is still some sort of a **God Class** since it is reponsible for both business logic (what needs to happen) and effects (changes to the system). This makes testing it cumbersome since the tests are more of integration tests which end up verifying both business logic and implementation details. 
+
+We need a scalable framework to build screens that lets us manage business logic and effects, and which is receptive to change.
+
+## Decision
+
+### Goals
+- Fulfill all the requirements of the [v2 architecture](./007-screen-architecture-v2.md#goals).
+- Separate business logic and effects.
+- Separate tests for business logic and effects so that we can manage them more easily.
+- Have a way to migrate from the v1 architecture to the v3 for existing screens.
+
+### Choosing a framework
+We evaluated many patterns and frameworks that are common in the industry, including but not limited to:
+
+- [Android recommended architecture](https://developer.android.com/jetpack/docs/guide)
+- [MvRx, by Airbnb](https://github.com/airbnb/MvRx)
+- [MVI](http://hannesdorfmann.com/android/mosby3-mvi-1)
+
+Eventually, we chose [Mobius](https://github.com/spotify/mobius), a reactive framework for managing state and side-effects as the basis for the v3 screen architecture. The [objectives](https://github.com/spotify/mobius/wiki/Objectives) of the framework also align very well with what we need from the architecture.
+
+#### Reference implementation
+A reference implementation of the complete architecture can be found at [this commit](https://github.com/simpledotorg/simple-android/tree/e51a6b2fad295f9da4064b95f51d95e3ab2f0e6c/app/src/main/java/org/simple/clinic/settings/changelanguage).
+
+#### Core components
+It is recommended to read the [Concepts](https://github.com/spotify/mobius/wiki/Concepts) page on the Mobius wiki before proceeding.
+
+We will take an example feature along with sample implementations for the various components and key classes:
+
+##### `Model`
+[**DOC**](https://github.com/spotify/mobius/wiki/Concepts#model)
```

Comment:
From what I understand of an ADR, it should talk about why we chose Mobius and how it solves for those reasons we have mentioned before. This document should be a justification for moving to Mobius.

I don't think we need to capture the implementation details and what do the different components look like and how does the architecture work, rather we should capture how those different components help us achieve what the old arch didn't.
Let's talk about this in person if you have doubts.

Reply:
These are all excellent points. My intention in capturing these implementation details is because we have some classes that we have written on top of Mobius to support our new architecture (like the `MobiusDelegate`, for example) and how to test them because we don't have any more documentation on this.

I think we can:
- Split the implementation details out into a separate doc which will live alongside the Mobius migration doc (so we'll have one doc for new screens and another for migrating older screens side-by-side)
- Update the ADR to only capture the decision and the other metrics you talked about.

What do you think?

Reply:
Yea. Sounds good. We can even have a small 30 min discussion tomorrow to make sure our understanding of an ADR is aligned. And I would also love for you to talk about this decision since you have the most context. We can pull-in Pratul so that he can share what writing ADRs mean. Wdyt?

Reply:
Works.

Reply:
I have made these changes. Can we review this again?